### ISSUE TITLE
Subscription missing messages

### ISSUE THREAD

**Issue by tylertreat-wf** [nats-io/jnats#5]
We've been hitting a bug which seems to be a result of a subscription "missing" a message. I know that NATS doesn't guarantee delivery, but this isn't something we've hit with the Go client and we hit it with jnats even just running gnatsd locally, making me think there's possibly a bug in the client. At first, I thought perhaps the message was arriving before the subscription was processed, but I added a flush to ensure this is not the case. Likewise, I tried using `request` and the issue still occurs.

The below test seems to reproduce the issue somewhat reliably:

```go
public class Main {

    public static void main(String[] args) throws Exception {
        ConnectionFactory cf = new ConnectionFactory(Constants.DEFAULT_URL);
        final Connection conn = cf.createConnection();
        conn.subscribe("foo", new MessageHandler() {
            @Override
            public void onMessage(Message message) {
                conn.publish(message.getReplyTo(), "hello".getBytes());
            }
        });
        for (int i = 0; i < 100; i++) {
            try {
                System.out.println(conn.request("foo", "test".getBytes(), 5000));
            } catch (TimeoutException e) {
                System.out.println("got timeout " + i);
                return;
            }
        }
    }

}
```

If you run this enough times, you should eventually hit a timeout:

```
{Subject=_INBOX.8DECFDBCE59AB390E13136A381;Reply=null;Payload=<hello>}
{Subject=_INBOX.B2EDC697845F196E9C04554226;Reply=null;Payload=<hello>}
got timeout 2
```

```
[64093] 2016/01/13 17:11:14.806572 [INF] Starting gnatsd version 0.7.3
[64093] 2016/01/13 17:11:14.806702 [INF] Listening for client connections on 0.0.0.0:4222
[64093] 2016/01/13 17:11:14.806766 [INF] gnatsd is ready
[64093] 2016/01/13 17:11:20.618532 [DBG] 127.0.0.1:64287 - cid:1 - Client connection created
[64093] 2016/01/13 17:11:20.625319 [TRC] 127.0.0.1:64287 - cid:1 - ->> [CONNECT {"verbose":false,"pedantic":false,"ssl_required":false,"name":"","lang":"java","version":"0.3.0-SNAPSHOT"}]
[64093] 2016/01/13 17:11:20.625394 [TRC] 127.0.0.1:64287 - cid:1 - ->> [PING]
[64093] 2016/01/13 17:11:20.625399 [TRC] 127.0.0.1:64287 - cid:1 - <<- [PONG]
[64093] 2016/01/13 17:11:20.635456 [TRC] 127.0.0.1:64287 - cid:1 - ->> [SUB foo  1]
[64093] 2016/01/13 17:11:20.636304 [TRC] 127.0.0.1:64287 - cid:1 - ->> [SUB _INBOX.8DECFDBCE59AB390E13136A381  2]
[64093] 2016/01/13 17:11:20.636325 [TRC] 127.0.0.1:64287 - cid:1 - ->> [UNSUB 2 1]
[64093] 2016/01/13 17:11:20.636336 [DBG] 127.0.0.1:64287 - cid:1 - Deferring actual UNSUB(_INBOX.8DECFDBCE59AB390E13136A381): 1 max, 0 received
[64093] 2016/01/13 17:11:20.636346 [TRC] 127.0.0.1:64287 - cid:1 - ->> [PUB foo _INBOX.8DECFDBCE59AB390E13136A381 4]
[64093] 2016/01/13 17:11:20.636354 [TRC] 127.0.0.1:64287 - cid:1 - ->> MSG_PAYLOAD: [test]
[64093] 2016/01/13 17:11:20.636383 [TRC] 127.0.0.1:64287 - cid:1 - <<- [MSG foo 1 _INBOX.8DECFDBCE59AB390E13136A381 4]
[64093] 2016/01/13 17:11:20.638169 [TRC] 127.0.0.1:64287 - cid:1 - ->> [PUB _INBOX.8DECFDBCE59AB390E13136A381 5]
[64093] 2016/01/13 17:11:20.638177 [TRC] 127.0.0.1:64287 - cid:1 - ->> MSG_PAYLOAD: [hello]
[64093] 2016/01/13 17:11:20.638185 [DBG] 127.0.0.1:64287 - cid:1 - Auto-unsubscribe limit of 1 reached for sid '2'
[64093] 2016/01/13 17:11:20.638193 [TRC] 127.0.0.1:64287 - cid:1 - <<- [MSG _INBOX.8DECFDBCE59AB390E13136A381 2 5]
[64093] 2016/01/13 17:11:20.645482 [TRC] 127.0.0.1:64287 - cid:1 - <-> [DELSUB 2]
[64093] 2016/01/13 17:11:20.646376 [TRC] 127.0.0.1:64287 - cid:1 - ->> [SUB _INBOX.B2EDC697845F196E9C04554226  3]
[64093] 2016/01/13 17:11:20.646397 [TRC] 127.0.0.1:64287 - cid:1 - ->> [UNSUB 3 1]
[64093] 2016/01/13 17:11:20.646403 [DBG] 127.0.0.1:64287 - cid:1 - Deferring actual UNSUB(_INBOX.B2EDC697845F196E9C04554226): 1 max, 0 received
[64093] 2016/01/13 17:11:20.646407 [TRC] 127.0.0.1:64287 - cid:1 - ->> [PUB foo _INBOX.B2EDC697845F196E9C04554226 4]
[64093] 2016/01/13 17:11:20.646411 [TRC] 127.0.0.1:64287 - cid:1 - ->> MSG_PAYLOAD: [test]
[64093] 2016/01/13 17:11:20.646416 [TRC] 127.0.0.1:64287 - cid:1 - <<- [MSG foo 1 _INBOX.B2EDC697845F196E9C04554226 4]
[64093] 2016/01/13 17:11:20.652036 [TRC] 127.0.0.1:64287 - cid:1 - ->> [PUB _INBOX.B2EDC697845F196E9C04554226 5]
[64093] 2016/01/13 17:11:20.652066 [TRC] 127.0.0.1:64287 - cid:1 - ->> MSG_PAYLOAD: [hello]
[64093] 2016/01/13 17:11:20.652088 [DBG] 127.0.0.1:64287 - cid:1 - Auto-unsubscribe limit of 1 reached for sid '3'
[64093] 2016/01/13 17:11:20.652106 [TRC] 127.0.0.1:64287 - cid:1 - <<- [MSG _INBOX.B2EDC697845F196E9C04554226 3 5]
[64093] 2016/01/13 17:11:20.652123 [TRC] 127.0.0.1:64287 - cid:1 - <-> [DELSUB 3]
[64093] 2016/01/13 17:11:25.658322 [TRC] 127.0.0.1:64287 - cid:1 - ->> [SUB _INBOX.472BDB7C069D4E3D178D78E927  4]
[64093] 2016/01/13 17:11:25.658357 [TRC] 127.0.0.1:64287 - cid:1 - ->> [UNSUB 4 1]
[64093] 2016/01/13 17:11:25.658368 [DBG] 127.0.0.1:64287 - cid:1 - Deferring actual UNSUB(_INBOX.472BDB7C069D4E3D178D78E927): 1 max, 0 received
[64093] 2016/01/13 17:11:25.658378 [TRC] 127.0.0.1:64287 - cid:1 - ->> [PUB foo _INBOX.472BDB7C069D4E3D178D78E927 4]
[64093] 2016/01/13 17:11:25.658385 [TRC] 127.0.0.1:64287 - cid:1 - ->> MSG_PAYLOAD: [test]
[64093] 2016/01/13 17:11:25.658395 [TRC] 127.0.0.1:64287 - cid:1 - <<- [MSG foo 1 _INBOX.472BDB7C069D4E3D178D78E927 4]
[64093] 2016/01/13 17:11:25.658405 [TRC] 127.0.0.1:64287 - cid:1 - ->> [UNSUB 4 0]
[64093] 2016/01/13 17:11:25.658412 [TRC] 127.0.0.1:64287 - cid:1 - <-> [DELSUB 4]
[64093] 2016/01/13 17:11:25.659005 [TRC] 127.0.0.1:64287 - cid:1 - ->> [PUB _INBOX.472BDB7C069D4E3D178D78E927 5]
[64093] 2016/01/13 17:11:25.664827 [TRC] 127.0.0.1:64287 - cid:1 - ->> MSG_PAYLOAD: [hello]
```

**Comment by mcqueary** [comment]
Hi Tyler, I’ll have a look. This was using the 0.3.0-SNAPSHOT, or the latest GitHub master? 

-- 
Larry McQueary
Sent with Airmail

On January 13, 2016 at 4:15:47 PM, Tyler Treat (notifications@github.com) wrote:

We've been hitting a bug which seems to be a result of a subscription "missing" a message. I know that NATS doesn't guarantee delivery, but this isn't something we've hit with the Go client and we hit it with jnats even just running gnatsd locally, making me think there's possibly a bug in the client. At first, I thought perhaps the message was arriving before the subscription was processed, but I added a flush to ensure this is not the case. Likewise, I tried using request and the issue still occurs.

The below test seems to reproduce the issue somewhat reliably:

    public static void main(String[] args) throws Exception {
        ConnectionFactory cf = new ConnectionFactory(Constants.DEFAULT_URL);
        final Connection conn = cf.createConnection();
        conn.subscribe("foo", new MessageHandler() {
            @Override
            public void onMessage(Message message) {
                conn.publish(message.getReplyTo(), "hello".getBytes());
            }
        });
        for (int i = 0; i < 100000; i++) {
            try {
                System.out.println(conn.request("foo", "test".getBytes(), 5000));
            } catch (TimeoutException e) {
                System.out.println("got timeout " + i);
                return;
            }
        }
    }
If you run this enough times, you should eventually hit a timeout:

{Subject=_INBOX.8DECFDBCE59AB390E13136A381;Reply=null;Payload=<hello>}
{Subject=_INBOX.B2EDC697845F196E9C04554226;Reply=null;Payload=<hello>}
got timeout 2

[64093] 2016/01/13 17:11:14.806572 [INF] Starting gnatsd version 0.7.3
[64093] 2016/01/13 17:11:14.806702 [INF] Listening for client connections on 0.0.0.0:4222
[64093] 2016/01/13 17:11:14.806766 [INF] gnatsd is ready
[64093] 2016/01/13 17:11:20.618532 [DBG] 127.0.0.1:64287 - cid:1 - Client connection created
[64093] 2016/01/13 17:11:20.625319 [TRC] 127.0.0.1:64287 - cid:1 - ->> [CONNECT {"verbose":false,"pedantic":false,"ssl_required":false,"name":"","lang":"java","version":"0.3.0-SNAPSHOT"}]
[64093] 2016/01/13 17:11:20.625394 [TRC] 127.0.0.1:64287 - cid:1 - ->> [PING]
[64093] 2016/01/13 17:11:20.625399 [TRC] 127.0.0.1:64287 - cid:1 - <<- [PONG]
[64093] 2016/01/13 17:11:20.635456 [TRC] 127.0.0.1:64287 - cid:1 - ->> [SUB foo  1]
[64093] 2016/01/13 17:11:20.636304 [TRC] 127.0.0.1:64287 - cid:1 - ->> [SUB _INBOX.8DECFDBCE59AB390E13136A381  2]
[64093] 2016/01/13 17:11:20.636325 [TRC] 127.0.0.1:64287 - cid:1 - ->> [UNSUB 2 1]
[64093] 2016/01/13 17:11:20.636336 [DBG] 127.0.0.1:64287 - cid:1 - Deferring actual UNSUB(_INBOX.8DECFDBCE59AB390E13136A381): 1 max, 0 received
[64093] 2016/01/13 17:11:20.636346 [TRC] 127.0.0.1:64287 - cid:1 - ->> [PUB foo _INBOX.8DECFDBCE59AB390E13136A381 4]
[64093] 2016/01/13 17:11:20.636354 [TRC] 127.0.0.1:64287 - cid:1 - ->> MSG_PAYLOAD: [test]
[64093] 2016/01/13 17:11:20.636383 [TRC] 127.0.0.1:64287 - cid:1 - <<- [MSG foo 1 _INBOX.8DECFDBCE59AB390E13136A381 4]
[64093] 2016/01/13 17:11:20.638169 [TRC] 127.0.0.1:64287 - cid:1 - ->> [PUB _INBOX.8DECFDBCE59AB390E13136A381 5]
[64093] 2016/01/13 17:11:20.638177 [TRC] 127.0.0.1:64287 - cid:1 - ->> MSG_PAYLOAD: [hello]
[64093] 2016/01/13 17:11:20.638185 [DBG] 127.0.0.1:64287 - cid:1 - Auto-unsubscribe limit of 1 reached for sid '2'
[64093] 2016/01/13 17:11:20.638193 [TRC] 127.0.0.1:64287 - cid:1 - <<- [MSG _INBOX.8DECFDBCE59AB390E13136A381 2 5]
[64093] 2016/01/13 17:11:20.645482 [TRC] 127.0.0.1:64287 - cid:1 - <-> [DELSUB 2]
[64093] 2016/01/13 17:11:20.646376 [TRC] 127.0.0.1:64287 - cid:1 - ->> [SUB _INBOX.B2EDC697845F196E9C04554226  3]
[64093] 2016/01/13 17:11:20.646397 [TRC] 127.0.0.1:64287 - cid:1 - ->> [UNSUB 3 1]
[64093] 2016/01/13 17:11:20.646403 [DBG] 127.0.0.1:64287 - cid:1 - Deferring actual UNSUB(_INBOX.B2EDC697845F196E9C04554226): 1 max, 0 received
[64093] 2016/01/13 17:11:20.646407 [TRC] 127.0.0.1:64287 - cid:1 - ->> [PUB foo _INBOX.B2EDC697845F196E9C04554226 4]
[64093] 2016/01/13 17:11:20.646411 [TRC] 127.0.0.1:64287 - cid:1 - ->> MSG_PAYLOAD: [test]
[64093] 2016/01/13 17:11:20.646416 [TRC] 127.0.0.1:64287 - cid:1 - <<- [MSG foo 1 _INBOX.B2EDC697845F196E9C04554226 4]
[64093] 2016/01/13 17:11:20.652036 [TRC] 127.0.0.1:64287 - cid:1 - ->> [PUB _INBOX.B2EDC697845F196E9C04554226 5]
[64093] 2016/01/13 17:11:20.652066 [TRC] 127.0.0.1:64287 - cid:1 - ->> MSG_PAYLOAD: [hello]
[64093] 2016/01/13 17:11:20.652088 [DBG] 127.0.0.1:64287 - cid:1 - Auto-unsubscribe limit of 1 reached for sid '3'
[64093] 2016/01/13 17:11:20.652106 [TRC] 127.0.0.1:64287 - cid:1 - <<- [MSG _INBOX.B2EDC697845F196E9C04554226 3 5]
[64093] 2016/01/13 17:11:20.652123 [TRC] 127.0.0.1:64287 - cid:1 - <-> [DELSUB 3]
[64093] 2016/01/13 17:11:25.658322 [TRC] 127.0.0.1:64287 - cid:1 - ->> [SUB _INBOX.472BDB7C069D4E3D178D78E927  4]
[64093] 2016/01/13 17:11:25.658357 [TRC] 127.0.0.1:64287 - cid:1 - ->> [UNSUB 4 1]
[64093] 2016/01/13 17:11:25.658368 [DBG] 127.0.0.1:64287 - cid:1 - Deferring actual UNSUB(_INBOX.472BDB7C069D4E3D178D78E927): 1 max, 0 received
[64093] 2016/01/13 17:11:25.658378 [TRC] 127.0.0.1:64287 - cid:1 - ->> [PUB foo _INBOX.472BDB7C069D4E3D178D78E927 4]
[64093] 2016/01/13 17:11:25.658385 [TRC] 127.0.0.1:64287 - cid:1 - ->> MSG_PAYLOAD: [test]
[64093] 2016/01/13 17:11:25.658395 [TRC] 127.0.0.1:64287 - cid:1 - <<- [MSG foo 1 _INBOX.472BDB7C069D4E3D178D78E927 4]
[64093] 2016/01/13 17:11:25.658405 [TRC] 127.0.0.1:64287 - cid:1 - ->> [UNSUB 4 0]
[64093] 2016/01/13 17:11:25.658412 [TRC] 127.0.0.1:64287 - cid:1 - <-> [DELSUB 4]
[64093] 2016/01/13 17:11:25.659005 [TRC] 127.0.0.1:64287 - cid:1 - ->> [PUB _INBOX.472BDB7C069D4E3D178D78E927 5]
[64093] 2016/01/13 17:11:25.664827 [TRC] 127.0.0.1:64287 - cid:1 - ->> MSG_PAYLOAD: [hello]

**Comment by tylertreat-wf** [comment]
0.3.0-SNAPSHOT. Thanks, Larry!

**Comment by mcqueary** [comment]
Easy fix. One line of code. The problem was one that I see frequently, as the parent of 3 boys: not flushing! Very different kind of output stream, however.

**Comment by mcqueary** [comment]
I deployed a new snapshot to Nexus earlier that incorporates this fix, based on the current `master`.

**Comment by tylertreat-wf** [comment]
@mcqueary Did you push a 0.3.1 snapshot? I only see 0.3.0-SNAPSHOT in Nexus.

**Comment by mcqueary** [comment]
I pushed it as 0.3.0. It should have a different timestamp in the full filename. I probably should have marked it as 0.3.1, but was in a hurry.

**Comment by tylertreat-wf** [comment]
We seem to still be hitting a similar problem:

```
    public static void main(String[] args) throws Exception {
        ConnectionFactory cf = new ConnectionFactory(Constants.DEFAULT_URL);
        final Connection conn = cf.createConnection();
        final Connection pub = cf.createConnection();
        for (int i = 0; i < 1000; i++) {
            Subscription sub = conn.subscribe("foo", "bar", new MessageHandler() {
                @Override
                public void onMessage(Message message) {
                    conn.publish(message.getReplyTo(), "hello".getBytes());
                }
            });
            conn.flush();
            try {
                System.out.println(pub.request("foo", "blah".getBytes(), 5000));
            } catch (TimeoutException e) {
                System.out.println("timeout" + i);
                return;
            }
            sub.unsubscribe();
        }
        System.out.println("done");
    }
```

Output:

```
{Subject=_INBOX.A928266065CB6C2F22DE4A703B;Reply=null;Payload=<hello>}
timeout1
```

```
[70939] 2016/01/15 09:41:59.748793 [INF] Starting gnatsd version 0.7.3
[70939] 2016/01/15 09:41:59.748937 [INF] Listening for client connections on 0.0.0.0:4222
[70939] 2016/01/15 09:41:59.749004 [INF] gnatsd is ready
[70939] 2016/01/15 09:42:29.391363 [DBG] 127.0.0.1:53486 - cid:1 - Client connection created
[70939] 2016/01/15 09:42:29.399728 [TRC] 127.0.0.1:53486 - cid:1 - ->> [CONNECT {"verbose":false,"pedantic":false,"ssl_required":false,"name":"","lang":"java","version":"0.3.0-SNAPSHOT"}]
[70939] 2016/01/15 09:42:29.399818 [TRC] 127.0.0.1:53486 - cid:1 - ->> [PING]
[70939] 2016/01/15 09:42:29.399825 [TRC] 127.0.0.1:53486 - cid:1 - <<- [PONG]
[70939] 2016/01/15 09:42:29.411909 [DBG] 127.0.0.1:53487 - cid:2 - Client connection created
[70939] 2016/01/15 09:42:29.412722 [TRC] 127.0.0.1:53487 - cid:2 - ->> [CONNECT {"verbose":false,"pedantic":false,"ssl_required":false,"name":"","lang":"java","version":"0.3.0-SNAPSHOT"}]
[70939] 2016/01/15 09:42:29.412760 [TRC] 127.0.0.1:53487 - cid:2 - ->> [PING]
[70939] 2016/01/15 09:42:29.412768 [TRC] 127.0.0.1:53487 - cid:2 - <<- [PONG]
[70939] 2016/01/15 09:42:29.416342 [TRC] 127.0.0.1:53486 - cid:1 - ->> [SUB foo bar 1]
[70939] 2016/01/15 09:42:29.416988 [TRC] 127.0.0.1:53487 - cid:2 - ->> [SUB _INBOX.A928266065CB6C2F22DE4A703B  1]
[70939] 2016/01/15 09:42:29.417211 [TRC] 127.0.0.1:53487 - cid:2 - ->> [UNSUB 1 1]
[70939] 2016/01/15 09:42:29.417220 [DBG] 127.0.0.1:53487 - cid:2 - Deferring actual UNSUB(_INBOX.A928266065CB6C2F22DE4A703B): 1 max, 0 received
[70939] 2016/01/15 09:42:29.417229 [TRC] 127.0.0.1:53487 - cid:2 - ->> [PUB foo _INBOX.A928266065CB6C2F22DE4A703B 4]
[70939] 2016/01/15 09:42:29.417236 [TRC] 127.0.0.1:53487 - cid:2 - ->> MSG_PAYLOAD: [blah]
[70939] 2016/01/15 09:42:29.417261 [TRC] 127.0.0.1:53486 - cid:1 - <<- [MSG foo 1 _INBOX.A928266065CB6C2F22DE4A703B 4]
[70939] 2016/01/15 09:42:29.419219 [TRC] 127.0.0.1:53486 - cid:1 - ->> [PUB _INBOX.A928266065CB6C2F22DE4A703B 5]
[70939] 2016/01/15 09:42:29.419230 [TRC] 127.0.0.1:53486 - cid:1 - ->> MSG_PAYLOAD: [hello]
[70939] 2016/01/15 09:42:29.419239 [DBG] 127.0.0.1:53486 - cid:1 - Auto-unsubscribe limit of 1 reached for sid '1'
[70939] 2016/01/15 09:42:29.419245 [TRC] 127.0.0.1:53487 - cid:2 - <<- [MSG _INBOX.A928266065CB6C2F22DE4A703B 1 5]
[70939] 2016/01/15 09:42:29.419251 [TRC] 127.0.0.1:53487 - cid:2 - <-> [DELSUB 1]
[70939] 2016/01/15 09:42:29.420546 [TRC] 127.0.0.1:53487 - cid:2 - ->> [SUB _INBOX.19C47390D5FE2BC559C0D40A5C  2]
[70939] 2016/01/15 09:42:29.420560 [TRC] 127.0.0.1:53487 - cid:2 - ->> [UNSUB 2 1]
[70939] 2016/01/15 09:42:29.420566 [DBG] 127.0.0.1:53487 - cid:2 - Deferring actual UNSUB(_INBOX.19C47390D5FE2BC559C0D40A5C): 1 max, 0 received
[70939] 2016/01/15 09:42:29.420570 [TRC] 127.0.0.1:53487 - cid:2 - ->> [PUB foo _INBOX.19C47390D5FE2BC559C0D40A5C 4]
[70939] 2016/01/15 09:42:29.420574 [TRC] 127.0.0.1:53487 - cid:2 - ->> MSG_PAYLOAD: [blah]
[70939] 2016/01/15 09:42:29.420581 [TRC] 127.0.0.1:53486 - cid:1 - <<- [MSG foo 1 _INBOX.19C47390D5FE2BC559C0D40A5C 4]
[70939] 2016/01/15 09:42:34.425826 [TRC] 127.0.0.1:53487 - cid:2 - ->> [UNSUB 2 0]
[70939] 2016/01/15 09:42:34.425851 [TRC] 127.0.0.1:53487 - cid:2 - <-> [DELSUB 2]
```

**Comment by mcqueary** [comment]
Yes, this seems to be a timing issue. I've been playing with it since last night. If I insert a 250 microsecond sleep in the `request()` method, I can get it to pass consistently for 10,000 messages. Anything less, and I see the intermittent timeouts you're seeing. 

It's just a convenience function. While I debug this, you could just replace it with the equivalent calls:

```java
	private Message _request(String subject, byte[] data, long timeout)
			throws TimeoutException, IOException 
	{
		String inbox 	= newInbox();
		Message m		= null;
		try (SyncSubscription s = subscribeSync(inbox, null))
		{
                        // Adjust sleep here as required
                        try { Thread.sleep(0, 250000); } catch (InterruptedException e) { /* NOOP */ }
			s.autoUnsubscribe(1);
			publish(subject, inbox, data);
			m = s.nextMessage(timeout);
		} catch (IOException | TimeoutException e) {
			throw(e);
		} 
		return m;		
	}
```

**Comment by kozlovic** [comment]
Larry, check your use of the flusher. I looks like you call kickFlusher() outside of the connection's lock. For the channel, you are using a LinkedBlockingQueue. This is not thread-safe by default. kickFlusher() checks the size of the queue to see if it should 'signal'. That could be a problem.

**Comment by tylertreat-wf** [comment]
> For the channel, you are using a LinkedBlockingQueue. This is not thread-safe by default. 

Is that true? I thought all `BlockingQueue` implementations were thread-safe?

**Comment by kozlovic** [comment]
Sorry, my mistake they are. But the problem is that although actions are atomic, in this case you would need still a lock. When checking if it needs to put a message in the queue (fc.getCount() == 0), this operation is atomic, and may return that there is something in the queue. By the time this function returns, the queue may be empty and sit in a wait. The caller will then decide that there was no need to add to the queue and the flusher won't do its job. Does that make sense?

**Comment by tylertreat-wf** [comment]
Yep, makes sense.

**Comment by kozlovic** [comment]
Conversely, it is possible that several threads add something to that channel. Say that the queue is empty. Several threads try to kickFlusher(). It is possible that all calls to getCount() return 0. They would then each put something in the queue. No big deal, though. In this case it is better to "over" notify, than miss a notification ;-)

**Comment by mcqueary** [comment]
Ivan, thanks, which call to kickFlusher are we talking about? There are many.

**Comment by kozlovic** [comment]
I see unsubscribe(), subscribe(), publish(). The fact that kickFlusher() checks on `bw != null` requires that it is called under the connection lock. But again, that would not be enough.

The flusher() thread is getting notifications outside of the connection's lock (and I understand why). But you would either have to use a lock to protect the signal and execution, or simply always add to the queue.

**Comment by mcqueary** [comment]
Always adding to the flusher queue seems like it would be very inefficient. I'll do some tests.

**Comment by kozlovic** [comment]
Agreed. That's not ideal. But try just to see if this is the issue at hand. You can even leave the kicFlusher() where they are for now (again, checking bw outside of the conn's lock is wrong, but let's ignore this for now).

**Comment by mcqueary** [comment]
Yep, I understand.

**Comment by mcqueary** [comment]
Thanks for pointing out the problems with `kickFlusher()`.  I am fixing them, but I don't believe that's the problem. Looking at the gnatsd log, I am finding that the issue seems to consistently be that the async message handler (responder) eventually does not receive a request message from the requestor, and therefore does not respond, which causes the timeout on the requestor side. So it's a problem with inbound messages on the async subscription.

**Comment by kozlovic** [comment]
If you look at Tyler's log of this morning, you see that the request mades it to the subscriber. It is the response that is not sent back, and then the requestor times-out after 5 seconds. So this would still be consistent with a buffer not flushed, wouldn't it?

**Comment by kozlovic** [comment]
To illustrate:

```
[70939] 2016/01/15 09:42:29.420574 [TRC] 127.0.0.1:53487 - cid:2 - ->> MSG_PAYLOAD: [blah]
[70939] 2016/01/15 09:42:29.420581 [TRC] 127.0.0.1:53486 - cid:1 - <<- [MSG foo 1 _INBOX.19C47390D5FE2BC559C0D40A5C 4]
[70939] 2016/01/15 09:42:34.425826 [TRC] 127.0.0.1:53487 - cid:2 - ->> [UNSUB 2 0]
[70939] 2016/01/15 09:42:34.425851 [TRC] 127.0.0.1:53487 - cid:2 - <-> [DELSUB 2]
```

cid2 is the publisher, cid1 is the subscriber. cid:1 receives the request 'blah' (size 4). After 5 seconds, the requestor cid2 times out and unsubscribes.

**Comment by mcqueary** [comment]
I'm looking at results from more recent code on my laptop, and over a single connection (based on Tyler's original example), so the signature is different. I'll get to the bottom of it. Thanks.

**Comment by ColinSullivan1** [comment]
Could there be an exception being thrown but ignored someplace?

I see a potential issue in the Channel with regarding the Object.wait() method.  Wait() can have spurious wake ups, which would result in undefined behavior in the channel.  I could see this causing problems and could be worth looking into here.  It should be fixed anyhow.

e.g. in Channel.get() - try { this.wait(); } catch (InterruptedException e) {} could cause get() to return when the channel is empty.

See:  http://docs.oracle.com/javase/6/docs/api/java/lang/Object.html#wait%28%29

**Comment by mcqueary** [comment]
I'm not using a `wait()` in the current Channel implementation. I removed all of that explicitly for this reason about a week ago :)

**Comment by mcqueary** [comment]
@kozlovic, Regardless of the different output, I think you're right. I see from tailing the log file along with my own tracing that somehow the request message isn't being kicked out to the server until the request timeout occurs.

**Comment by kozlovic** [comment]
Because when the timeout occurs, the unsubscribe() will send a proto that will flush the buffer. That being said, this is not what we see in the logfile that I excerpted above. The request is really sent to the client 5 seconds before the timeout. Anyway, I will let you investigate. We are trying to help, but I know this can be distracting. Good luck and let us know if you need anything!

**Comment by mcqueary** [comment]
Appreciate the help always!

**Comment by mcqueary** [comment]
@tylertreat-wf Thanks for your patience. I'm going to push a 0.3.1-SNAPSHOT in a little while that fixes the flushing problem, hopefully for good! (and thanks @kozlovic )

**Comment by tylertreat-wf** [comment]
No worries! Thanks for digging into this. :)

**Comment by mcqueary** [comment]
Tyler, I think we have resolved this in [jnats-0.3.1](http://search.maven.org/#artifactdetails%7Cio.nats%7Cjnats%7C0.3.1%7C). Please try it out.

```
<dependency>
    <groupId>io.nats</groupId>
    <artifactId>jnats</artifactId>
    <version>0.3.1</version>
</dependency>
```
